lib/concurrent-limiter.js

```diff
@@ -0,0 +1,92 @@
+const Signal = require('signal-promise')
+
+class ConcurrentLimiterError extends Error {
+  static NEVER_PROMISE = new Promise(() => {})
+
+  constructor (msg, code, fn = ConcurrentLimiterError) {
+    super(`${code}: ${msg}`)
+    this.code = code
+
+    if (Error.captureStackTrace) {
+      Error.captureStackTrace(this, fn)
+    }
+  }
+
+  get name () {
+    return 'ConcurrentLimiterError'
+  }
+
+  static CONCURRENT_LIMITER_DESTROYED () {
+    return new ConcurrentLimiterError(
+      'The concurrent limiter is destroyed',
+      'CONCURRENT_LIMITER_DESTROYED',
+      ConcurrentLimiterError.CONCURRENT_LIMITER_DESTROYED
+    )
+  }
+}
+
+module.exports = class ConcurrentLimiter {
+  /**
+   * @param {object} options
+   * @param {number} options.maxConcurrent - Maximum concurrent executions.
+   */
+  constructor ({ maxConcurrent } = {}) {
+    this._maxConcurrent = maxConcurrent
+    this._active = 0
+    this._releaseSignal = new Signal()
+    this._destroyed = false
+  }
+
+  _tryAcquire () {
+    if (this._active < this._maxConcurrent) {
+      this._active++
+      return true
+    }
+
+    return false
+  }
+
+  _release () {
+    this._active--
+    this._releaseSignal.notify()
+  }
+
+  /**
+   * Execute an async function with a timeout.
+   *
+   * @template T
+   * @param {() => Promise<T>} fn
+   * @param {object} [options] - Options for the execution.
+   * @param {Promise<void>} [options.abortSignalPromise] - Promise that resolves when the execution should be aborted.
+   * @returns {Promise<T>}
+   */
+  async execute (fn, { abortSignalPromise = ConcurrentLimiterError.NEVER_PROMISE } = {}) {
```

Comment:
(to discuss) In terms of API design, I wonder if we should make the abortSignal a non-optional parameter. It's never really a good idea to queue requests without any abort logic, and we avoid having to define a default